warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java

```diff
@@ -0,0 +1,431 @@
+    public Object visit(ASTAndNode node, Object data) {
+        if (!isValid)
+            return data;
+        
+        QueryPropertyMarker.Instance instance = QueryPropertyMarker.findInstance(node);
+        if (instance.isAnyType()) {
+            // descend into the marker's source node
+            JexlNode source = instance.getSource();
+            source.childrenAccept(this, data);
+            return data;
+        } else if (isAnyChildAWrappedSingleTerm(node)) {
+            // invalidate
+            invalidate(node, "wrapped single child");
+            return data;
+        } else {
+            // otherwise continue descent
+            node.childrenAccept(this, data);
+            return data;
+        }
+    }
+    
+    /*
+     * Helper methods
+     */
+    
+    // is this node a parentheses
+    private boolean isParen(JexlNode node) {
+        return node instanceof ASTReference && node.jjtGetNumChildren() == 1 && node.jjtGetChild(0) instanceof ASTReferenceExpression;
+    }
+    
+    // fetch the grandchild
+    private JexlNode grandChild(JexlNode node) {
+        return node.jjtGetChild(0).jjtGetChild(0);
+    }
+    
+    private boolean isAnyChildAWrappedSingleTerm(JexlNode node) {
+        for (int i = 0; i < node.jjtGetNumChildren(); i++) {
+            JexlNode child = node.jjtGetChild(i);
+            if (isWrappedSingleTerm(child)) {
+                return true;
+            }
+        }
+        return false;
+    }
+    
+    private boolean isWrappedSingleTerm(JexlNode node) {
+        if (isParen(node)) {
+            boolean isMarkerNode = QueryPropertyMarker.findInstance(node).isAnyType();
+            if (!isMarkerNode) {
+                JexlNode unwrapped = JexlASTHelper.dereference(node);
+                
+                return !(unwrapped instanceof ASTAndNode || unwrapped instanceof ASTOrNode);
+            }
+        }
+        return false;
+    }
+    
+    private void invalidate(JexlNode node, String cause) {
+        isValid = false;
+        log.info("Invalid node (" + cause + "): " + JexlStringBuildingVisitor.buildQueryWithoutParse(node));
+    }
+    
+    /*
+     * Pass through
+     */
+    
+    @Override
+    public Object visit(ASTJexlScript node, Object data) {
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    @Override
+    public Object visit(SimpleNode node, Object data) {
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTReferenceExpression node, Object data) {
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    // descend into negated branches of the query
+    @Override
+    public Object visit(ASTNotNode node, Object data) {
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    /*
+     * Short circuits
+     */
+    
+    @Override
+    public Object visit(ASTBlock node, Object data) {
+        return data;
```

Comment:
It looks like we could let these delegate to the parent implementations as well. Though here we're short circuiting I guess. Is that the intent? Are these parts of JEXL we know we don't utilize? 

Not saying we do it here, but if there is a subset of the language we're not supporting, we might bubble these "no-ops" up into our BaseVisitor.

Reply:
Oh, just saw the comment above. Missed that initially.